x-pack/plugins/security_solution/public/detections/components/rules/rule_preview/index.tsx

```diff
@@ -228,10 +247,27 @@ const RulePreviewComponent: React.FC<RulePreviewProps> = ({
 
   const { startTransaction } = useStartTransaction();
 
+  const [isRefreshing, setIsRefreshing] = useState(false);
+  useEffect(() => {
+    if (!isRefreshing) {
+      return;
+    }
+    createPreview();
+    setIsRefreshing(false);
+  }, [isRefreshing, createPreview]);
```

Comment:
Why do you need `useState` + `useEffect` here? Could we call `createPreview()` in the click handler directly?

Reply:
In order to call preview without updating data, we need to update `usePreviewRoute` hook which is dependant on advanced options: `timeframeStart` and `timeframeEnd`. That is why we refresh timeframe on preview button click in `handlePreviewClick`. This triggers updating of the `usePreviewRoute` hook which responsible for actual preview results fetching. Only, after that we can call `createPreview` which we get from the updated `usePreviewRoute` hook.

So, the steps are:
1. We click preview button which triggers the `handlePreviewClick` handler
2. We update timeframe to update the `usePreviewRoute` hook
3. We get updated `createPreview` method and preview route state is ready to handle preview request
4. We trigger `createPreview` by utilising the `isRefreshing` state and `useEffect`

Hope this makes sense! :-)

Reply:
Yeah, that logic is a little bit vague. So the data fetching doesn't happen imperatively, but instead, a series of state changes and reactions in `useEffect` hooks trigger it.

I would suggest using `react-query` to fetch preview data to reduce the logic involved in state manipulations and potentially get rid of many of the `useEffect` hooks, but this is definitely out of the scope of the PR.